File: leetcode-java/src/main/java/chapter03/BinaryTreeConnectNext.java
package chapter03;

import utils.TreeNode;

import java.util.LinkedList;
import java.util.Queue;

/**
 * <p>1. 填充每个节点的下一个右侧节点指针</p>
 * <p>2. 填充每个节点的下一个右侧节点指针 II</p>
 * <h2>填充每个结点的 Next 指针</h2>
 * <p>1. 层序遍历非常简单</p>
 * <p>2. 深度遍历需要使用已经建立好的 Next 指针</p>
 * <p>两个问题</p>
 * <p>1. 树为满二叉树的时候, 填充 Next 指针</p>
 * <p>2. 树为普通二叉树的时候, 填充 Next 指针</p>
 * <p>总结: 填充指针这种类型的题目, 通常都可以考虑在填充之后立刻使用这些指针</p>
 */
public class BinaryTreeConnectNext {

    /**
     * 填充每个节点的下一个右侧节点指针: 满二叉树
     */
    public static class FullBinaryTree {

        /**
         * 层序遍历, 广度优先
         */
        public static TreeNode connectNextBFS(TreeNode root) {
            TreeNode previous = null;
            TreeNode currentLevelEnd = root;
            TreeNode nextLevelEnd = null;
            Queue<TreeNode> queue = new LinkedList<>();
            if (root != null) {
                queue.offer(root);
            }
            while (!queue.isEmpty()) {
                TreeNode node = queue.poll();
                if (previous != null) {
                    previous.next = node;
                }
                previous = node;
                if (node.left != null) {
                    queue.offer(node.left);
                    nextLevelEnd = node.left;
                }
                if (node.right != null) {
                    queue.offer(node.right);
                    nextLevelEnd = node.right;
                }
                if (currentLevelEnd == root) {
                    previous = null;
                    currentLevelEnd = nextLevelEnd;
                }
            }
            return root;
        }

        /**
         * 类前序遍历, 深度优先, 空间复杂度在不计压栈空间时可以认为是 O(1)
         */
        public static TreeNode connectNextDFS(TreeNode root) {
            connectNextDFS(root, null);
            return root;
        }

        // 思路: 记录自己的相邻节点, 然后进行连接
        public static void connectNextDFS(TreeNode root, TreeNode neighbor) {
            if (root == null) {
                return;
            }
            root.next = neighbor;
            connectNextDFS(root.left, root.right);
            // 相邻节点的左子结点就是下一层节点的相邻节点
            connectNextDFS(root.right, neighbor != null ? neighbor.left : null);
        }

        /**
         * 类层序遍历, 广度优先: 真正意义上空间复杂度为 O(1)
         */
        public static TreeNode connectNextLoop(TreeNode root) {
            if (root == null) {
                return null;
            }
            TreeNode current = root;
            while (current.left != null) {
                // 将当前父节点作为链表的头结点
                TreeNode head = current;
                // 遍历链表
                while (head != null) {
                    // 将父节点的左子结点的 next 指向父节点的右子节点
                    head.left.next = head.right;
                    // 如果父节点有 next, 那么就可以将父节点的右子节点的 next 指向相邻节点的左子结点
                    if (head.next != null) {
                        head.right.next = head.next.left;
                    }
                    head = head.next;
                }
                // 从下一层的最左节点继续
                current = current.left;
            }
            return root;
        }

    }

    /**
     * 填充每个节点的下一个右侧节点指针 II: 普通二叉树
     */
    public static class BinaryTree {

        /**
         * 层序遍历, 广度优先, 和满二叉树完全相同的做法
         */
        public static TreeNode connectNextBFS(TreeNode root) {
            TreeNode previous = null;
            TreeNode currentLevelEnd = root;
            TreeNode nextLevelEnd = null;
            Queue<TreeNode> queue = new LinkedList<>();
            if (root != null) {
                queue.offer(root);
            }
            while (!queue.isEmpty()) {
                TreeNode node = queue.poll();
                if (previous != null) {
                    previous.next = node;
                }
                previous = node;
                if (node.left != null) {
                    queue.offer(node.left);
                    nextLevelEnd = node.left;
                }
                if (node.right != null) {
                    queue.offer(node.right);
                    nextLevelEnd = node.right;
                }
                if (currentLevelEnd == root) {
                    previous = null;
                    currentLevelEnd = nextLevelEnd;
                }
            }
            return root;
        }

        /**
         * 层序遍历, 广度优先, 真正意义上空间复杂度为 O(1)
         */
        public static TreeNode connectNextLoop(TreeNode root) {
            if (root == null) {
                return null;
            }
            TreeNode current = root;
            while (current != null) {
                TreeNode head = current;
                // 记录前驱节点的目的, 因为需要向后遍历链表找到有子节点的父节点, 所以需要记录前驱节点, 从而连接起来
                // 记录后继节点的目的, 因为不清楚父节点有没有子节点, 所以记录下一层从哪里开始, 是从左子结点开始还是右子节点
                TreeNode previous = null, next = null;
                while (head != null) {
                    if (head.left != null) {
                        if (previous != null) {
                            previous.next = head.left;
                        }
                        if (next == null) {
                            next = head;
                        }
                        previous = head.left;
                    }
                    if (head.right != null) {
                        if (previous != null) {
                            previous.next = head.right;
                        }
                        if (next == null) {
                            next = head;
                        }
                        previous = head.right;
                    }
                    head = head.next;
                }
                current = next;
            }
            return root;
        }

    }

}

File: leetcode-go/chapter03/binary_tree_connect_next.go
package chapter03

import (
	"github.com/emirpasic/gods/queues/linkedlistqueue"
	nodes "leetcode-go/node"
)

// 填充每个节点的下一个右侧节点指针
type FullBinaryTreeConnectNext struct {}

// 广度优先, 迭代实现, 空间复杂度为 O(1)
func (tree *FullBinaryTreeConnectNext) Connect(root *nodes.TreeNode) *nodes.TreeNode {
	if root == nil {
		return nil
	}
	current := root
	for current.Left != nil {
		head := current
		for head != nil {
			head.Left.Next = head.Right
			if head.Next != nil {
				head.Right.Next = head.Next.Left
			}
			head = head.Next
		}
		current = current.Left
	}
	return root
}

// 层序遍历, 广度优先，迭代实现
func (tree *FullBinaryTreeConnectNext) ConnectBfs(root *nodes.TreeNode) *nodes.TreeNode {
	var currentLevelEnd *nodes.TreeNode = root
	var previous, nextLevelEnd *nodes.TreeNode = nil, nil
	queue := linkedlistqueue.New()
	if root != nil {
		queue.Enqueue(root)
	}
	for !queue.Empty() {
		element, _ := queue.Dequeue()
		node := element.(*nodes.TreeNode)
		if previous != nil {
			previous.Next = node
		}
		previous = node
		if node.Left != nil {
			queue.Enqueue(node.Left)
			nextLevelEnd = node.Left
		}
		if node.Right != nil {
			queue.Enqueue(node.Right)
			nextLevelEnd = node.Right
		}
		if currentLevelEnd == node {
			previous = nil
			currentLevelEnd = nextLevelEnd
		}
	}
	return root
}

// 前序遍历, 深度优先, 递归实现
func (tree *FullBinaryTreeConnectNext) ConnectDfs(root *nodes.TreeNode) *nodes.TreeNode {
	var recur func(current, neighbor *nodes.TreeNode)
	recur = func(current, neighbor *nodes.TreeNode) {
		if current == nil {
			return
		}
		current.Next = neighbor
		if neighbor != nil {
			neighbor = neighbor.Left
		}
		recur(current.Left, current.Right)
		recur(current.Right, neighbor)
	}
	recur(root, nil)
	return root
}


// 填充每个节点的下一个右侧节点指针 II
type BinaryTreeConnectNext struct {}

// 广度优先, 迭代实现, 空间复杂度为 O(1)
func (tree *BinaryTreeConnectNext) Connect(root *nodes.TreeNode) *nodes.TreeNode {
	if root == nil {
		return nil
	}
	current := root
	for current != nil {
		var head *nodes.TreeNode = current
		var previous, next *nodes.TreeNode = nil, nil
		for head != nil {
			if head.Left != nil {
				if previous != nil {
					previous.Next = head.Left
				}
				if next == nil {
					next = head.Left
				}
				previous = head.Left
			}
			if head.Right != nil {
				if previous != nil {
					previous.Next = head.Right
				}
				if next == nil {
					next = head.Right
				}
				previous = head.Right
			}
			head = head.Next
		}
		current = next
	}
	return root
}

// 层序遍历, 广度优先, 迭代实现
func (tree *BinaryTreeConnectNext) ConnectBfs(root *nodes.TreeNode) *nodes.TreeNode {
	var currentLevelEnd *nodes.TreeNode = root
	var previous, nextLevelEnd *nodes.TreeNode = nil, nil
	queue := linkedlistqueue.New()
	if root != nil {
		queue.Enqueue(root)
	}
	for !queue.Empty() {
		element, _ := queue.Dequeue()
		node := element.(*nodes.TreeNode)
		if previous != nil {
			previous.Next = node
		}
		previous = node
		if node.Left != nil {
			queue.Enqueue(node.Left)
			nextLevelEnd = node.Left
		}
		if node.Right != nil {
			queue.Enqueue(node.Right)
			nextLevelEnd = node.Right
		}
		if currentLevelEnd == node {
			previous = nil
			currentLevelEnd = nextLevelEnd
		}
	}
	return root
}